### PR TITLE
Update Exchange Rate API

### PR DESCRIPTION
**Pull request by tammytee**
## Summary

This PR will
- update the value of the `api_url` config variable. It appears that the Exchange Rates API is now hosted on the [API Layer](https://apilayer.com/marketplace/exchangerates_data-api) marketplace
- simplify the call to the Exchange Rates API inside the `RequestBuilder` class by using the `HTTP` facade provided by Laravel
- fix failing tests and migrate the `phpunit.xml` file to the updated format

### Failing Tests

Upon running the test suite, the following tests failed due to discrepancies in float precision using `assertEquals`:
- `tests/Unit/ConvertBetweenDateRangeTest.php`
- `tests/Unit/ConvertTest.php`

**Error example**
Screen Shot 2022-11-03 at 9 27 09 AM

To fix this, I updated the places where these values are calculated inside the `ExchangeRate.php` class to round the values to the expected precision. If this is not the desired way to resolve this issue, you may alternatively update your test assertion to `assertEqualsWithDelta` like so:

```php
// tests/Unit/ConvertBetweenDateRangeTest.php

/** @test */
public function converted_values_between_date_range_are_returned_and_exchange_rates_are_not_cached()
{
    // ...
    
    $this->assertEqualsWithDelta([
            '2019-11-08' => 116.06583254,
            '2019-11-06' => 116.23446817,
            '2019-11-07' => 115.68450522,
            '2019-11-05' => 116.12648497,
            '2019-11-04' => 115.78362356,
    ], $currencies, 0.000000001);
}
```

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -67,7 +67,7 @@ The package has been developed and tested to work with the following minimum req
 ## Getting Your API Key
 
 As of 1st April 2021, the exchangeratesapi.io now requires an API key to use the service. To get an API key, head over to
-[https://exchangeratesapi.io/pricing](https://exchangeratesapi.io/pricing). You can sign up for free or use the paid tiers.
+[https://apilayer.com/marketplace/exchangerates_data-api#pricing](https://apilayer.com/marketplace/exchangerates_data-api#pricing). You can sign up for free or use the paid tiers.
 
 Please note that at the time of writing this, you will need to be on at least the 'Basic' plan to make request via HTTPS. You
 will also be required to have at least the 'Professional' plan to use the ` convertBetweenDateRange() ` and ` exchangeRateBetweenDateRange() `
```
Comment (TammyTee) [inline README.md:73]: Not sure if the information noted here regarding plan restrictions is still accurate now that the API is moved. Needs additional research.

**Comment by ash-jc-allen** [comment]
Hey @TammyTee! Thanks for the PR 😄

I've not been keeping up to date with the API so my knowledge of it is a little fuzzy at the moment. Does the original API still work, or has it been completely abandoned?

**Comment by TammyTee** [comment]
Hi @ash-jc-allen 👋🏾  - this is a good point I didn't consider. It does appear that the [old API is still being maintained](https://exchangeratesapi.io/documentation/). Although, all of the call to actions on their website to get a new API key will redirect users to the API Layer marketplace. The API Layer marketplace keys do not work with the original API:

Screen Shot 2022-11-03 at 10 40 51 AM

**Comment by ash-jc-allen** [comment]
Hey @TammyTee! Sorry, I've not been ignoring this. I've been trying to think about different ways of supporting the new API and can't quite decide which would be the easiest to go for. The ideas that I've had so far are:

- Do a major version upgrade of the package and drop support for the old API completely. This seems a bit harsh though and might cause a lot of people to stop using the package.
- Add support for the new API and use something (like your config/env toggle) to determine which API to use. We could then deprecate support for the legacy API and completely drop support in the future.
- Add support for the new API and pass the `access_token` query param (so that it works with the legacy API) _and_ pass the `apiKey` header (so that it works with the newer API). I'm not sure if I like this idea, but I feel like it would be the easiest to implement with the least amount of resistance for upgrading.

I'd love to hear your thoughts on this because I can't quite make my mind up about which approach to take. If you have any other suggestions, I'm all ears.

I'm gonna share this PR on Twitter too so that I can get some more feedback/suggestions from the community too on how to tackle this 😄

**Comment by peterfox** [comment]
Now seeing the code clear, I think the best thing is to make two versions of the request builder, the builder itself implementing a contract, one to use the API key in the header, the other to use it as a query parameter instead, then make a factory to make it possible to switch implementations back and forth, defaulting with the legacy until a major version release.

**Comment by jaetoole** [comment]
@ash-jc-allen I like @peterfox's approach. It's hard because the old API is so coupled into the codebase. Hindsight; should have used DTOs so we could have switched APIs and as long as we can hydrate the DTOs we would be okay.

If we're going down the major version update, I think we should add in some DTOs maybe.

**Comment by ash-jc-allen** [comment]
Hey @peterfox and @jaetoole! Huge thanks for both of your inputs, I really appreciate it.

Yeah I 100% agree that using more abstraction in the codebase could potentially be beneficial here. This was the very first package I ever made, so there are definitely a few things that I'd do differently if I was to do it again from scratch.

At the moment, I'm leaning towards maybe passing the API key as a query param and also in a header. This wouldn't be a permanent fix, but could probably help to avoid a breaking change for the time being and support both APIs. I could then have a bit more time to implement something like the manager pattern to put a more permanent solution in place. I guess this would also open up the package even more and allow it to support completely different exchange rate API services too (but I'm not too sure about this at the moment).

What are everyone's thoughts on this sort of approach? 😄

**Comment by jaetoole** [comment]
@ash-jc-allen I have a similar situation where I use the repository pattern to hydrate DTOs from different Clients which all implement a Client contract. It's an API with multiple versions which they like to break. This probably would be over complex for this use case. You almost want a default client to pass into a builder which implements a Client interface, then add the ability to pass in a custom client maybe? Not really given it tons of thought!

**Comment by peterfox** [comment]
> This was the very first package I ever made, so there are definitely a few things that I'd do differently if I was to do it again from scratch.

To be fair, it's a mistake in their API as well really as the key itself is a piece of sensitive info so it shouldn't have been sent in the query parameters ever. Even when using HTTPS the URL isn't encrypted so the key is exposed in network traffic.

> At the moment, I'm leaning towards maybe passing the API key as a query param and also in a header. This wouldn't be a permanent fix, but could probably help to avoid a breaking change for the time being and support both APIs. 

I don't really see changing the RequestBuilder to a contract or abstract as a breaking change. As long as the new package is installed and still behaves the same, that's the best you can hope for sometimes. As I mentioned changing internals is breaking for people who have extended the functionality but in all honesty, you can't be responsible for such users. My approach has been if people extend the internals of a package they should be prepared for some issues. Most of the time these get flagged easily because an interface definition has changed or a class no longer exists.

> What are everyone's thoughts on this sort of approach? 😄

I think in the long term it wouldn't be a bad idea to just make the package allow for multiple providers. I can't see there being many providers though that are as extensive as the current one that allows for the same functionality to be done in one API call. Either way that is very much a breaking change. If you're going to go down that route, I'd open up discussions for other potential breaking changes current users might want. 

I'm happy to try and make a new PR for my suggested short-term fix at some point this week to explain thoughts on what could change if you like. Don't think it's a crazy amount of work required to do so.

**Comment by ash-jc-allen** [comment]
Hey @TammyTee! I'm just dropping a quick message on here to let you know that I've not forgotten about this PR. Life's super busy at the moment, but I'm hoping to look at this properly next week (if possible) 🙂

**Comment by TammyTee** [comment]
Sounds good @ash-jc-allen 🙂. Thanks for the update!

**Comment by ash-jc-allen** [comment]
Hey @tammytee! I'm really sorry for taking this long to add support for the new API. I've finally managed to get v6 of the package ready.

It has support for the existing API and the new API. It's also using the concept of "drivers" so other APIs can also easily be supported.

Assuming that I don't find any oddities in the latest changes, I'm planning on releasing v6 at some point this week. So I'll close this PR.

But a huge thanks for pushing me to make these changes though, I really appreciate it! 😄

v6 PR: https://github.com/ash-jc-allen/laravel-exchange-rates/pull/91

**Pull request closed without merging**